### DISCUSSION TITLE
Convert to webp using imagick

### DISCUSSION THREAD

**Discussion by Alerinos** [SixLabors/ImageSharp.Web#156] [Q&A] [answered]
How to add a new format? I use imagick which converts my format to webp + creates compression.
```
var small = new MagickImage(GET FILE DIR);
small.Resize("GET &width", "GET &height");
small.Format = MagickFormat.WebP;
small.Quality = GET &quality;
small.Write($"GET CACHE NAME");
```
How to execute this code to save itself as cache?
Or how is your work progress on the webp?

**Comment by KimPiks** [comment]
I have the same problem

**Comment by JimBobSquarePants** [comment, marked as answer]
You need to implement your own `IImageDecoder` and `IImageEncoder` instances and register them against the configuration. 

See the reference codec from our ImageSharp unit tests for examples. 

https://github.com/SixLabors/ImageSharp/tree/dc0982f1864b0c41fcc1a79853b3d16746906835/tests/ImageSharp.Tests/TestUtilities/ReferenceCodecs

**Comment by Alerinos** [comment]
May I have more details or some example of how to register this in asp.net? I'm not that experienced on asp.net, thank you so much!

**Reply by JimBobSquarePants** [reply]
You register it against the Configuration class instance in startup. There’s an example application containing startup examples in this repository.

**Reply by Alerinos** [reply]
Are you talking about it? https://github.com/SixLabors/ImageSharp.Web/tree/master/samples

**Reply by JimBobSquarePants** [reply]
Yes

**Reply by Alerinos** [reply]
```
services.AddImageSharp()
                .AddProcessor<ImageSharpWebP>();
```
```
    public class ImageSharpWebP : IImageEncoder, IImageWebProcessor
    {
        private readonly System.Drawing.Imaging.ImageFormat imageFormat;

        public ImageSharpWebP(ImageFormat imageFormat)
        {
            this.imageFormat = imageFormat;
        }

        public IEnumerable<string> Commands => throw new NotImplementedException();

        public void Encode<TPixel>(Image<TPixel> image, Stream stream) where TPixel : unmanaged, IPixel<TPixel>
        {
            throw new NotImplementedException();
        }

        public Task EncodeAsync<TPixel>(Image<TPixel> image, Stream stream, CancellationToken cancellationToken) where TPixel : unmanaged, IPixel<TPixel>
        {
            throw new NotImplementedException();
        }

        public FormattedImage Process(
            FormattedImage image,
            ILogger logger,
            IDictionary<string, string> commands,
            CommandParser parser,
            CultureInfo culture)
        {
            ResizeOptions options = new ResizeOptions {};

            if (options != null)
            {
                image.Image.Mutate(x => x.Resize(options));
            }

            return image;
        }

    }
```
``
 System.AggregateException: „Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: SixLabors.ImageSharp.Web.Processors.IImageWebProcessor Lifetime: Singleton ImplementationType: net_core_www.Core.Lib.ImageSharpWebP': Unable to resolve service for type 'System.Drawing.Imaging.ImageFormat' while attempting to activate 'net_core_www.Core.Lib.ImageSharpWebP'.)”
 ``
 
 What am I doing wrong?

**Reply by JimBobSquarePants** [reply]
You’re not following my advice not samples I’ve linked to.

You need to register the encoder with the Configuration class. Instead you’ve incorrectly registered it as a processor. You’re also injecting an unresolvable System.Drawing ImageFormat instance into the constructor for some reason (The error message is explicitly telling you this). ImageSharp has nothing to do with System.Drawing so I do not understand why you would be doing that. 

I suggest you take your time to a study the API and complementary docs and the code so that you understand what you are working with.

**Reply by JimBobSquarePants** [reply]
@Alerinos Looks like someone has already written a wrapper for a webp encoder and published it on NuGet. You could use that until we release ours.

https://www.nuget.org/packages/Shorthand.ImageSharp.WebP/

**Reply by Alerinos** [reply]
Thanks for your help, I'm already testing. Do you need to register for this to work?

**Reply by JimBobSquarePants** [reply]
You need to register any encoder with the Configuration class as I described above. See the reference codecs I linked to in the ImageSharp unit tests for  an example.
